piptools/resolver.py

```diff
@@ -139,23 +148,41 @@ def _group_constraints(self, constraints):
             django~=1.5,<1.9,>=1.4.2
             flask~=0.7
 
+        If any of the constraints are a link to be installed, that constraint will
+        be returned, as long as it satisifies all the other constraints on the
+        same package. If it doesn't, or if there are multiple linked constraints
+        that are different, then this will rais an error.
+
         """
-        for _, ireqs in full_groupby(constraints, key=_dep_key):
+        for name, ireqs in full_groupby(constraints, key=_dep_key):
             ireqs = list(ireqs)
             editable_ireq = first(ireqs, key=lambda ireq: ireq.editable)
             if editable_ireq:
                 yield editable_ireq  # ignore all the other specs: the editable one is the one that counts
                 continue
 
-            ireqs = iter(ireqs)
-            combined_ireq = next(ireqs)
-            combined_ireq.comes_from = None
-            for ireq in ireqs:
-                # NOTE we may be losing some info on dropped reqs here
-                combined_ireq.req.specifier &= ireq.req.specifier
-                # Return a sorted, de-duped tuple of extras
-                combined_ireq.extras = tuple(sorted(set(tuple(combined_ireq.extras) + tuple(ireq.extras))))
-            yield combined_ireq
+            links = set(ireq.link for ireq in ireqs if ireq.link is not None)
+            if len(links) > 1:
+                raise IncompatibleRequirements(
+                    first(ireqs, key=lambda ireq: ireq.link == links.pop()),
```

Comment:
Is there a way to write this in a more straightforward manner? Popping off the links list feels to me like there might be a bug in that the links set is not ordered, so there might be a potential for this 1st call to the 'first' method could be looking for the link associated with the 2nd ireq. And then the next line won't find the other ireq because its link value has already been popped off the list of links.

E.g.
```
>>> from first import first
>>> links = set(['bar', 'foo'])
>>> first(['bar', 'foo'], key=lambda x: x == links.pop())
>>> links
set([])
```